python313Packages.mlx: 0.21.1 -> 0.26.3

cameronyule opened the pull request:
Fixing broken build (https://github.com/NixOS/nixpkgs/issues/349991), ensuring automated tests are being checked as part of the build, and bumping version to latest.

Closes #349991

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

cameronyule committed:
maintainers: add cameronyule

cameronyule committed:
python313Packages.mlx: 0.21.1 -> 0.25.2

Fixes failing build, adds testing, upgrades to latest version.

Co-authored-by: Sarah Clark <seclark@nextquestion.net>

cameronyule commented:
### `nixpkgs-review` result

Generated using [nixpkgs-review](https://github.com/Mic92/nixpkgs-review).

---

### `aarch64-darwin`

<details>
<summary>✅ 4 packages built:</summary>

python312Packages.mlx
python312Packages.mlx.dist
python313Packages.mlx
python313Packages.mlx.dist
</details>

sarahec reviewed (changes requested):
Mostly small changes to catch up with current practice.

sarahec commented on pkgs/development/python-modules/mlx/default.nix:
```diff
@@ -1,94 +1,164 @@
+    rev = "8fa6eb65236618e28fd7710a0fba565f7faa1848";
+    hash = "sha256-15FvyPOFqTOr5vdWQoPnZz+mYH919++EtghjozDlnSA=";
   };
   nlohmann_json = fetchFromGitHub {
     owner = "nlohmann";
     repo = "json";
     rev = "v3.11.3";
     hash = "sha256-7F0Jon+1oWL7uqet5i1IgHX0fUw/+z0QwEcA3zs5xHg=";
   };
```
```suggestion
```

Since all we need is the source, and there's a matching package in nixpkgs, we can just use the package.

sarahec commented on pkgs/development/python-modules/mlx/default.nix:
```diff
@@ -1,94 +1,164 @@
+  cmake,
+
+  # buildInputs
+  apple-sdk_14,
```
```suggestion
  apple-sdk_14,
  fmt_10,
```

sarahec commented on pkgs/development/python-modules/mlx/default.nix:
```diff
@@ -1,94 +1,164 @@
+        # or by changing the upstream (e.g., https://github.com/zed-industries/zed/discussions/7016).
+        (lib.cmakeBool "MLX_BUILD_METAL" false)
+        (lib.cmakeOptionType "filepath" "FETCHCONTENT_SOURCE_DIR_GGUFLIB" "${gguf-tools}")
+        (lib.cmakeOptionType "filepath" "FETCHCONTENT_SOURCE_DIR_JSON" "${nlohmann_json}")
+        (lib.cmakeOptionType "filepath" "FETCHCONTENT_SOURCE_DIR_FMT" "${fmt}")
```
```suggestion
        (lib.cmakeOptionType "filepath" "FETCHCONTENT_SOURCE_DIR_JSON" "${nlohmann_json.src}")
        (lib.cmakeOptionType "filepath" "FETCHCONTENT_SOURCE_DIR_FMT" "${fmt_10.src}")
```

sarahec commented on pkgs/development/python-modules/mlx/default.nix:
```diff
@@ -1,94 +1,164 @@
+    rev = "10.2.1";
+    hash = "sha256-pEltGLAHLZ3xypD/Ur4dWPWJ9BGVXwqQyKcDWVmC3co=";
   };
+in
+let
+  mlx = buildPythonPackage rec {
```
```suggestion
  mlx = buildPythonPackage rec {
```

sarahec commented:
This is fantastic, thank you so much! Welcome to the maintainer team.

I'm sorry this wasn't picked up sooner. Two changes to start with:
1. Could you rebase these atop current `master` please? Some big Darwin changes landed in that time.
2. Release 0.26.1 just came out (probably for Apple's WWDC). Let's update to it.

I also removed github checkouts where we have the package in nixpkgs (this includes the package sources), fixed a redundant `let...in`, and fixed `with lib;` in `meta`. I tested these on a copy of your code and they all work.

sarahec commented:
Closes #349991

cameronyule committed:
python313Packages.mlx: 0.25.2 -> 0.26.3

cameronyule commented:
Thanks @sarahec! Just confirming that I've:

1. Rebased atop master
2. Incorporated the review feedback
3. Bumped to the latest version of mlx (v0.26.3)
4. Built the package successfully locally, on aarch64-darwin
5. Confirmed the mlx tests still pass
6. Confirmed nixpkgs-review still passes

Just shout if you have any further feedback :)

sarahec commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 410209`
Commit: `cfe1e8cfb87745d454a58a1f8a23fe4b9bd4f066`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python312Packages.mlx</li>
    <li>python312Packages.mlx.dist</li>
    <li>python313Packages.mlx</li>
    <li>python313Packages.mlx.dist</li>
  </ul>
</details>

sarahec commented:
You're a rock star. Thank you.

sarahec reviewed (approved):
LGTM! 

(Looks Good to Me, in case you haven't seen this before)

phaer reviewed (approved):
LGTM, could build on aarch64-darwin.

Thanks you! :tada:
